funsor/sum_product.py

```diff
@@ -138,36 +141,151 @@ def sequential_sum_product(sum_op, prod_op, trans, time, step):
+        if isinstance(step, dict):
+            step = frozenset(step.items())
+        if step_names is None:
+            step_names = frozenset((k, k) for pair in step for k in pair)
+        if isinstance(step_names, dict):
+            step_names = frozenset(step_names.items())
+        return super().__call__(sum_op, prod_op, trans, time, step, step_names)
+
+
+class MarkovProduct(Funsor, metaclass=MarkovProductMeta):
+    """
+    Lazy representation of :func:`sequential_sum_product` .
+
+    :param AssociativeOp sum_op: A marginalization op.
+    :param AssociativeOp prod_op: A Bayesian fusion op.
+    :param Funsor trans: A sequence of transition factors,
+        usually varying along the ``time`` input.
+    :param time: A time dimension.
+    :type time: str or Variable
+    :param dict step: A str-to-str mapping of "previous" inputs of ``trans``
+        to "current" inputs of ``trans``.
+    :param dict step_names: Optional, for internal use by alpha conversion.
+    """
+    def __init__(self, sum_op, prod_op, trans, time, step, step_names):
+        assert isinstance(sum_op, AssociativeOp)
+        assert isinstance(prod_op, AssociativeOp)
+        assert isinstance(trans, Funsor)
+        assert isinstance(time, Variable)
+        assert isinstance(step, frozenset)
+        assert isinstance(step_names, frozenset)
+        step = dict(step)
+        step_names = dict(step_names)
+        assert all(isinstance(k, str) for k in step_names.keys())
+        assert all(isinstance(v, str) for v in step_names.values())
+        assert set(step_names) == set(step).union(step.values())
+        inputs = OrderedDict((step_names.get(k, k), v)
+                             for k, v in trans.inputs.items()
+                             if k != time.name)
+        output = trans.output
+        fresh = frozenset(step_names.values())
+        bound = frozenset(step_names.keys()) | {time.name}
+        super().__init__(inputs, output, fresh, bound)
+        self.sum_op = sum_op
+        self.prod_op = prod_op
+        self.trans = trans
+        self.time = time
+        self.step = step
+        self.step_names = step_names
+
+    def _alpha_convert(self, alpha_subs):
+        assert self.bound.issuperset(alpha_subs)
+        time = Variable(alpha_subs.get(self.time.name, self.time.name),
+                        self.time.output)
+        step = frozenset((alpha_subs.get(k, k), alpha_subs.get(v, v))
+                         for k, v in self.step.items())
+        step_names = frozenset((alpha_subs.get(k, k), v)
+                               for k, v in self.step_names.items())
+        alpha_subs = {k: to_funsor(v, self.trans.inputs[k])
+                      for k, v in alpha_subs.items()
+                      if k in self.trans.inputs}
+        trans = substitute(self.trans, alpha_subs)
+        return self.sum_op, self.prod_op, trans, time, step, step_names
+
+    def eager_subs(self, subs):
```

Comment:
Is `MarkovProduct.eager_subs` exercised by the existing HMM distribution tests?  If not can you add an `eager_subs` test to `test_sum_product.py`?

Reply:
`MarkovProduct.eager_subs` is exercised by tests/examples/test_bart.py